CDCgov/prime-reportstream
# platform/jimfuqian/12929-Create-YAML-schema-for-settings

This PR:

Implement #12929 

Create settings schemas (json / yaml schemas) to be used to validate settings, e.g. organizations.yml, and settings on the deployed ENVs
Enforce the settings validation using the schemas

Initial settings'(YAML) json schema generation:

Use the online tool:

https://codebeautify.org/yaml-to-json-schema-generator

Paste in the local organizations.yml (seems sufficient with regard to fields / sub fields varieties) or settings download from STAGING, and obtain the initial json schema (in one piece), then hand tweak the auto generated schema:

e.g.
- Change additionalProperties: from false to true, be lenient on minor sub fields unless there is need to be restrict
- Break up schema into:
  - settings.json
  - organization.json
  - receiver.json
  - sender.json
 
For json/yaml schemas validation, check out: [JSON Schema](https://json-schema.org/)

Test Steps:

1. Pass all local tests: unit tests, integration tests, smoke tests
2. Perform the manual tests (listed below)

# Manual tests:

## Manual test: Run CLI validate-setting like this:

```bash
./prime validate-setting -i settings/organizations.yml -s settings/schemas/settings.json
```

Should see happy case validation report like:

validation completed: validation messages count: 0

Could also tweak the organizations.yml, e.g. add a field 'countyName' for an organization whose 'jurisdiction' is FEDERAL, and run the above command line, it should report error:

`validation completed: validation messages count: 1
error: $[0]: For FEDERAL jurisdiction, countyName must NOT present
schema path: $.items.$ref.$ref.then.allOf[1].not`

## Manual test: get the staging settings and validate it with settings schemas:

```bash
./prime login --env staging
./prime multiple-settings get -o organizations.staging.02.15.2024.yml --env staging
./prime validate-setting -i organizations.staging.02.15.2024.yml -s settings/schemas/settings.json
```
Note:

The validation should report 0 error, and a bonus is that the download organizations.staging.yml become smaller (due to trimming off the fields with null / empty, see below:

-rw-r--r--  1 jimfuqian-mac-pro  staff  438871 Feb  6 13:57 /Users/jimfuqian-mac-pro/prime-reportstream/prime-router/organizations.staging.02.06.2023.yml
-rw-r--r--   1 jimfuqian-mac-pro  staff  315608 Feb 20 16:01 organizations.staging.02.20.2024.yml

Size of settings yml become 315608 from 438871.


## Manual test: get local organization and do a diff
 
```bash
./prime organization get -n waters -o org_waters.yml
./prime organization diff -i org_waters.yml
```
this should report 0 diff

## Manual test: get local organization 'waters' and validate
 
```bash
./prime organization get -n waters -o org_waters.yml
./prime validate-setting -i org_waters.yml -s settings/schemas/organization.json
```
this should report 0 error
Note, the keys part is not checked by schemas, because it is an attachment from APiKeys API call.

## Manual testing from a local FE:

1. start frontend on local setup - see frontend README
2. login into FE
3. pick a organization to edit - say: ignore with FEDERAL as jurisdiction, put a value e.g. "Santa Clara" into field "County Name", and "Save", you will see a validation error from newly introduced json schema:

![image](https://github.com/CDCgov/prime-reportstream/assets/135686833/90b7e059-036b-4cb2-be8d-3f5fc634f3da)

## Manual test: add a receiver to waters organization:

1. login to localhost:3000 (yarn run dev from FE folder)
2. ![image](https://github.com/CDCgov/prime-reportstream/assets/135686833/2edcbc94-284b-4642-880e-e744b05a8e07)
3. New receiver added by e.g. clone a existing receiver and change the name, and other fields values, remember to pick time zone, and Check syntax and "Save".
4. 
![image](https://github.com/CDCgov/prime-reportstream/assets/135686833/3343773c-fdb0-46f3-8652-c6c8b7d88607)

5. Can also set fields with inconsistent values, and should be able to see validation error(s) at the "Save" time..., shown below, adding mappedConditionFilter to receiver with topic 'covid-19', got validation error from Settings API:
![image](https://github.com/CDCgov/prime-reportstream/assets/135686833/52f3c895-cc9c-4308-a776-e51b2ab9f45a)

Click "Save":

![image](https://github.com/CDCgov/prime-reportstream/assets/135686833/ff0fd39e-aae6-41fa-bba8-717352b85b84)

## Changes
- Created json (yaml) schemas for organizations.yml (structured)
   - settings/schemas/settings.json
   - settings/schemas/organization.json
   - settings/schemas/sender.json
   - settings/schemas/receiver.json 
- In the json schemas: created rules that insure the consistency of the settings:
   - e.g. for organization: when 'jurisdiction': FEDERAL, 'stateCode' and 'countyName' should not present
   - e.g. for receiver: when 'topic' is legacy, e.g. covid-19, 'conditionFilter' and 'mappedConditionFilter' should not present 
- Above mentioned rules also serve as example for future rules
- Introduced prime CLI to perform settings validation: ./prime validate-settings --help (does not need server running) 
- Changed instance (Organization, Sender, Receiver) validation in SettingsFacade.kt to use json schema validation
- Removed validation code that perform the instance validation (now done through json schema validation) 
- Json schema validator used: [Json Schema Validator](https://github.com/networknt/json-schema-validator)

## Checklist

### Testing

### Process

## Linked Issues
- Fixes #issue

## To Be Done
- #13458

## Specific Security-related subjects a reviewer should pay specific attention to
- Does this PR introduce new endpoints?
    - new endpoint A
    - new endpoint B
- Does this PR include changes in authentication and/or authorization of existing endpoints?
- Does this change introduce new dependencies that need vetting?
- Does this change require changes to our infrastructure?
- Does logging contain sensitive data?
- Does this PR include or remove any sensitive information itself?

If you answered '_yes_' to any of the questions above, conduct a detailed Review that addresses at least:

- What are the potential security threats and mitigations? Please list the _STRIDE_ threats and how they are mitigated
    - **S**poofing (faking authenticity)
        - Threat _T_, which could be achieved by _A_, is mitigated by _M_
    - **T**ampering (influence or sabotage the integrity of information, data, or system)
    - **R**epudiation (the ability to dispute the origin or originator of an action)
    - **I**nformation disclosure (data made available to entities who should not have it)
    - **D**enial of service (make a resource unavailable)
    - **E**levation of Privilege (reduce restrictions that apply or gain privileges one should not have)
- Have you ensured logging does not contain sensitive data?
- Have you received any additional approvals needed for this change?
